template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/CmTemplateProcessor.java

```diff
@@ -243,7 +247,31 @@ private Map<String, Set<ServiceComponent>> getNonGatewayServicesByHostGroup() {
                                 .filter(i -> !i.getComponent().equalsIgnoreCase("GATEWAY"))
                                 .collect(toSet())
                 ));
+    }
+
+    public List<String> getUnhealthyDependentHostGroups(StackDto stackdto, Set<String> dependentComponent, String undefinedDependency) {
+        Set<String> dependentHostGroupNames = getNonGatewayComponentsByHostGroup().entrySet().stream()
+                .filter(hgComponents ->
+                        isDependentComponentPresentInHostGroup(hgComponents.getValue(), dependentComponent))
+                .map(Entry::getKey)
+                .collect(toSet());
+        List<String> allUnHealthyHostGroupNames = stackdto.getAllAvailableInstances().stream()
+                .filter(instanceMetaData -> !InstanceStatus.SERVICES_HEALTHY.equals(instanceMetaData.getInstanceStatus()))
+                .map(InstanceMetadataView::getInstanceGroupName)
+                .collect(toList());
```

Comment:
toSet? We dont'n want an entry per host.

Reply:
Okay, I will change list to set.